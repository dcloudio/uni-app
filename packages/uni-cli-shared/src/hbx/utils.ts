import path from 'path'
import { once } from '@dcloudio/uni-shared'

export const isInHBuilderX = once(() => {
  // 自动化测试传入了 HX_APP_ROOT(其实就是UNI_HBUILDERX_PLUGINS)
  if (process.env.HX_APP_ROOT) {
    process.env.UNI_HBUILDERX_PLUGINS = process.env.HX_APP_ROOT + '/plugins'
    return true
  }
  try {
    const { name } = require(path.resolve(
      process.cwd(),
      '../about/package.json'
    ))
    if (name === 'about') {
      process.env.UNI_HBUILDERX_PLUGINS = path.resolve(process.cwd(), '..')
      return true
    }
  } catch (e) {
    // console.error(e)
  }
  return false
})
