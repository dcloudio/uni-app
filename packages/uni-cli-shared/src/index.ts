export * from './fs'
export * from './mp'
export * from './url'
export * from './env'
export * from './hbx'
export * from './ssr'
export * from './vue'
export * from './uts'
export * from './logs'
export * from './i18n'
export * from './deps'
export * from './json'
export * from './vite'
export * from './utils'
export * from './easycom'
export * from './constants'
export * from './preprocess'
export * from './postcss'
export * from './filter'
export * from './esbuild'
export * from './resolve'
export * from './scripts'
export * from './platform'
export * from './utsUtils'

export {
  parseUniExtApi,
  parseUniExtApis,
  parseInjects,
  Define,
  DefineOptions,
  Defines,
  getUniExtApiProviderRegisters,
  formatExtApiProviderName,
} from './uni_modules'
export {
  parseUniModulesArtifacts,
  resolveEncryptUniModule,
} from './uni_modules.cloud'
import type { EncryptPackageJson } from './uni_modules.cloud'
export type EncryptArtifacts = EncryptPackageJson['uni_modules']['artifacts']

export { M } from './messages'

export * from './exports'
export { checkUpdate } from './checkUpdate'
