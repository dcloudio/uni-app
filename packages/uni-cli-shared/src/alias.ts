import { initModuleAlias } from './hbx/alias'

initModuleAlias()
