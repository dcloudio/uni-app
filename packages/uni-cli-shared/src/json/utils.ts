import fs from 'fs'
import path from 'path'
import { isString } from '@vue/shared'
import type { CompilerError } from '@vue/compiler-core'
import {
  type Node,
  type ParseError,
  parseTree,
  printParseErrorCode,
} from 'jsonc-parser'

import { normalizePath } from '../utils'

import { offsetToLineColumn } from '../vite/plugins/vitejs/utils'

interface CheckPagesJsonError extends CompilerError {
  offsetStart: number
  offsetEnd: number
}

export function checkPagesJson(jsonStr: string, inputDir: string) {
  if (!inputDir) {
    return false
  }
  const errors: ParseError[] = []
  const root = parseTree(jsonStr, errors)
  if (!root) {
    if (errors.length) {
      for (const error of errors) {
        const { line, column } = offsetToLineColumn(jsonStr, error.offset)
        throw {
          name: 'SyntaxError',
          code: error.error,
          message: printParseErrorCode(error.error),
          loc: {
            start: { line, column },
          },
          offsetStart: error.offset,
          offsetEnd: error.offset + error.length,
        } as CheckPagesJsonError
      }
    }
    return false
  }
  const pagePathNodes = walkNodes(findRootNode(root, ['pages']))
  findRootNode(root, ['subPackages', 'subpackages']).forEach((node) => {
    const subPackageRoot = findSubPackageRoot(node)
    if (subPackageRoot) {
      findRootNode(node, ['pages']).forEach((subNode) => {
        walkNodes(subNode.children ?? []).forEach((node) => {
          pagePathNodes.push({
            ...node,
            value: normalizePath(path.join(subPackageRoot, node.value)),
          })
        })
      })
    }
  })

  if (pagePathNodes.length) {
    for (const node of pagePathNodes) {
      const pagePath: string = node.value ?? ''
      if (pageExistsWithCaseSync(path.join(inputDir, pagePath))) {
        continue
      }
      const { line, column } = offsetToLineColumn(jsonStr, node.offset)
      throw {
        name: 'CompilerError',
        code: 'CompilerError',
        message: `The page path "${pagePath}" does not exist`,
        loc: {
          start: { line, column },
        },
        offsetStart: node.offset,
        offsetEnd: node.offset + node.length,
      } as unknown as CheckPagesJsonError
    }
  }
  return true
}

function pageExistsWithCaseSync(pagePath: string) {
  try {
    const files = fs.readdirSync(path.dirname(pagePath))
    const basename = path.basename(pagePath)
    const uvuePage = basename + '.uvue'
    const vuePage = basename + '.vue'
    return files.some((file) => file === uvuePage || file === vuePage)
  } catch (e) {
    return false
  }
}

function findSubPackageRoot(node: Node) {
  const child = node.children?.find(
    (child) =>
      child.type === 'property' &&
      child.children &&
      child.children.find(
        (child) => child.type === 'string' && child.value === 'root'
      )
  )
  if (child && child.children?.length === 2) {
    return child.children[1].value
  }
  return ''
}

function findRootNode(node: Node, property: string[]) {
  const { type, children } = node
  if (type === 'object' && children) {
    const child = children.find(
      (child) =>
        child.type === 'property' &&
        child.children &&
        child.children.find(
          (child) => child.type === 'string' && property.includes(child.value)
        )
    )
    if (child) {
      const node = child.children!.find((child) => child.type === 'array')
      return node?.children ?? []
    }
  }
  return []
}

function walkNodes(node: Node[]) {
  const pagePathNodes: Node[] = []
  node.forEach((node) => walkNode(node, pagePathNodes))
  return pagePathNodes
}

function walkNode(node: Node, pagePathNodes: Node[]) {
  const { type, children } = node
  if (type === 'property' && children && children.length === 2) {
    const maybePagePathNode = children[0]
    const maybePagePathValueNode = children[1]
    if (
      maybePagePathNode.type === 'string' &&
      maybePagePathNode.value === 'path' &&
      maybePagePathValueNode.type === 'string' &&
      isString(maybePagePathValueNode.value)
    ) {
      pagePathNodes.push(maybePagePathValueNode)
    }
  }
  if (children) {
    children.forEach((node) => walkNode(node, pagePathNodes))
  }
}
