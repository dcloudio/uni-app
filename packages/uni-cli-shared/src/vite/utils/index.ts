import { isNormalCompileTarget } from '../../utils'
import type { ResolvedConfig } from 'vite'

export * from './ast'
export * from './url'
export * from './plugin'
export * from './utils'

// 内置组件css列表，h5平台需要合并进去首页css中
export const buildInCssSet = new Set<string>()

export function isCombineBuiltInCss(config: ResolvedConfig) {
  if (!isNormalCompileTarget()) {
    return false
  }
  return config.command === 'build' && config.build.cssCodeSplit
}
