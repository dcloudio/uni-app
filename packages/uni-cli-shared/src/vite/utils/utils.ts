import type { ConfigEnv, ResolvedConfig, UserConfig } from 'vite'
import type { RollupError } from 'rollup'
import type { CompilerError } from '@vue/compiler-sfc'
import { extend, hasOwn } from '@vue/shared'
import { codeFrameColumns } from '@babel/code-frame'
import { offsetToStartAndEnd } from '../plugins/vitejs/utils'
import { enableSourceMap } from '../../utils'

export function withSourcemap(config: ResolvedConfig) {
  if (!process.env.UNI_APP_SOURCEMAP) {
    if (config.build && hasOwn(config.build, 'sourcemap')) {
      if (!!config.build.sourcemap) {
        process.env.UNI_APP_SOURCEMAP = 'true'
      } else {
        // vite 的 build 模式默认是false，而非web端的dev也是用build模式，所以不能这样判断
        // process.env.UNI_APP_SOURCEMAP = 'false'
      }
    }
  }
  return enableSourceMap()
}

export function isInHybridNVue(config: UserConfig | ResolvedConfig): boolean {
  return (config as any).nvue && process.env.UNI_RENDERER !== 'native'
}

export function isSsr(
  command: ConfigEnv['command'],
  config: UserConfig | ResolvedConfig
) {
  if (command === 'serve') {
    return !!(config.server && config.server.middlewareMode)
  }
  if (command === 'build') {
    return !!(config.build && config.build.ssr)
  }
  return false
}

export function createRollupError(
  plugin: string,
  id: string,
  error: CompilerError | SyntaxError,
  source?: string
): RollupError {
  const { message, name, stack } = error
  const rollupError: RollupError = extend(new Error(message), {
    id,
    plugin,
    name,
    stack,
  })

  if ('code' in error && error.loc) {
    rollupError.loc = {
      file: id,
      line: error.loc.start.line,
      column: error.loc.start.column,
    }
    if (source && source.length > 0) {
      if ('offsetStart' in error && 'offsetEnd' in error) {
        rollupError.frame = codeFrameColumns(
          source,
          offsetToStartAndEnd(
            source,
            error.offsetStart as number,
            error.offsetEnd as number
          )
        )
      } else {
        rollupError.frame = codeFrameColumns(source, error.loc)
      }
    }
  }
  if (id) {
    // 指定了id后，不让后续的rollup重写
    Object.defineProperty(rollupError, 'id', {
      get() {
        return id
      },
      set(_v) {},
    })
  }
  return rollupError
}

export const generateCodeFrameColumns = codeFrameColumns
