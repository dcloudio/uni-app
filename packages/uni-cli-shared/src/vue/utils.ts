import { camelize, extend, isString } from '@vue/shared'
import { isComponentTag } from '@dcloudio/uni-shared'
import {
  type AttributeNode,
  type ComponentNode,
  type DirectiveNode,
  type ElementNode,
  ElementTypes,
  type ExpressionNode,
  NodeTypes,
  type Position,
  type RootNode,
  type SourceLocation,
  type TemplateChildNode,
  type TransformContext,
  createSimpleExpression,
  isCoreComponent,
  isStaticExp,
  locStub,
} from '@vue/compiler-core'
import { createAssetUrlTransformWithOptions } from './transforms/templateTransformAssetUrl'
import { createSrcsetTransformWithOptions } from './transforms/templateTransformSrcset'
import { isDirectiveNode } from '../vite/utils/ast'
import { parseVueRequest } from '../vite/utils/url'
import { EXTNAME_VUE_RE } from '../constants'

export const VUE_REF = 'r'
export const VUE_REF_IN_FOR = 'r-i-f'

export function isVueSfcFile(id: string) {
  const { filename, query } = parseVueRequest(id)
  return EXTNAME_VUE_RE.test(filename) && !query.vue
}

export function isUserComponent(
  node: RootNode | TemplateChildNode,
  context: {
    isBuiltInComponent: TransformContext['isBuiltInComponent']
  }
): node is ComponentNode {
  return (
    node.type === NodeTypes.ELEMENT &&
    node.tagType === ElementTypes.COMPONENT &&
    !isComponentTag(node.tag) &&
    !isCoreComponent(node.tag) &&
    !context.isBuiltInComponent(node.tag)
  )
}

export function createAttributeNode(
  name: string,
  content: string
): AttributeNode {
  return {
    type: NodeTypes.ATTRIBUTE,
    loc: locStub,
    nameLoc: locStub,
    name,
    value: {
      type: NodeTypes.TEXT,
      loc: locStub,
      content,
    },
  }
}

function createClassAttribute(clazz: string): AttributeNode {
  return createAttributeNode('class', clazz)
}

export function addStaticClass(node: ElementNode, clazz: string) {
  const classProp = node.props.find(
    (prop) => prop.type === NodeTypes.ATTRIBUTE && prop.name === 'class'
  ) as AttributeNode | undefined

  if (!classProp) {
    return node.props.unshift(createClassAttribute(clazz))
  }

  if (classProp.value) {
    return (classProp.value.content = classProp.value.content + ' ' + clazz)
  }
  classProp.value = {
    type: NodeTypes.TEXT,
    loc: locStub,
    content: clazz,
  }
}

export function createDirectiveNode(
  name: string,
  arg: string,
  exp?: string | ExpressionNode
): DirectiveNode {
  return {
    type: NodeTypes.DIRECTIVE,
    name,
    modifiers: [],
    loc: locStub,
    arg: createSimpleExpression(arg, true),
    exp: isString(exp) ? createSimpleExpression(exp, false) : exp,
  }
}

export function createOnDirectiveNode(name: string, value: string) {
  return createDirectiveNode('on', name, value)
}

export function createBindDirectiveNode(
  name: string,
  value: string | ExpressionNode
) {
  return createDirectiveNode('bind', name, value)
}

export function createUniVueTransformAssetUrls(base: string) {
  return {
    base,
    includeAbsolute: true,
    tags: {
      audio: ['src'],
      video: ['src', 'poster'],
      img: ['src'],
      image: ['src'],
      'cover-image': ['src'],
      // h5
      'v-uni-audio': ['src'],
      'v-uni-video': ['src', 'poster'],
      'v-uni-image': ['src'],
      'v-uni-cover-image': ['src'],
      // nvue
      'u-image': ['src'],
      'u-video': ['src', 'poster'],
    },
  }
}

export function getBaseNodeTransforms(base: string) {
  const transformAssetUrls = createUniVueTransformAssetUrls(base)
  return [
    createAssetUrlTransformWithOptions(transformAssetUrls),
    createSrcsetTransformWithOptions(transformAssetUrls),
  ]
}

export function renameProp(name: string, prop?: DirectiveNode | AttributeNode) {
  if (!prop) {
    return
  }
  if (isDirectiveNode(prop)) {
    if (prop.arg && isStaticExp(prop.arg)) {
      prop.arg.content = name
    }
  } else {
    prop.name = name
  }
}

export function isPropNameEquals(
  prop: AttributeNode | DirectiveNode,
  name: string
): boolean {
  if (prop.type === NodeTypes.ATTRIBUTE) {
    const propName = camelize(prop.name)
    return propName === name
  } else if (prop.type === NodeTypes.DIRECTIVE && prop.rawName) {
    const propName = camelize(prop.rawName.slice(1))
    return propName === name
  }
  return false
}

// @vue/compiler-core 没有导出 getLoc，先使用旧版本的 getInnerRange
export function getInnerRange(
  loc: SourceLocation,
  offset: number,
  length: number
): SourceLocation {
  const source = loc.source.slice(offset, offset + length)
  const newLoc: SourceLocation = {
    source,
    start: advancePositionWithClone(loc.start, loc.source, offset),
    end: loc.end,
  }

  if (length != null) {
    newLoc.end = advancePositionWithClone(
      loc.start,
      loc.source,
      offset + length
    )
  }

  return newLoc
}

export function advancePositionWithClone(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  return advancePositionWithMutation(
    extend({}, pos),
    source,
    numberOfCharacters
  )
}

// advance by mutation without cloning (for performance reasons), since this
// gets called a lot in the parser
export function advancePositionWithMutation(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  let linesCount = 0
  let lastNewLinePos = -1
  for (let i = 0; i < numberOfCharacters; i++) {
    if (source.charCodeAt(i) === 10 /* newline char code */) {
      linesCount++
      lastNewLinePos = i
    }
  }

  pos.offset += numberOfCharacters
  pos.line += linesCount
  pos.column =
    lastNewLinePos === -1
      ? pos.column + numberOfCharacters
      : numberOfCharacters - lastNewLinePos

  return pos
}
