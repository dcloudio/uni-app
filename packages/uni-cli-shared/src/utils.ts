import fs from 'fs'
import os from 'os'
import path from 'path'
import colors from 'picocolors'
import { camelize, capitalize } from '@vue/shared'
export { default as hash } from 'hash-sum'
import {
  EXTNAME_TS_RE,
  PAGE_EXTNAME,
  PAGE_EXTNAME_APP,
  X_PAGE_EXTNAME,
  X_PAGE_EXTNAME_APP,
} from './constants'

import {
  type ElementNode,
  NodeTypes,
  type RootNode,
  type TemplateChildNode,
} from '@vue/compiler-core'
import type { ParserPlugin } from '@babel/parser'
import { getPlatformDir } from './platform'
import { isInHBuilderX } from './hbx'
import { parseManifestJsonOnce } from './json'

// 专为 uts.ts 服务
export { camelize, capitalize, isArray } from '@vue/shared'

// export let isRunningWithYarnPnp: boolean
// try {
//   isRunningWithYarnPnp = Boolean(require('pnpapi'))
// } catch {}

export const isWindows = os.platform() === 'win32'
export function normalizePath(id: string): string {
  return isWindows ? id.replace(/\\/g, '/') : id
}

export function checkElementNodeTag(
  node: RootNode | TemplateChildNode | null | undefined,
  tag: string
): node is ElementNode {
  return !!node && node.type === NodeTypes.ELEMENT && node.tag === tag
}

/**
 * 根据 path 返回合法 js 变量
 * @param str pages.json.page.path
 * @returns
 */
export function normalizeIdentifier(str: string) {
  let _str = str.replace(/[^a-zA-Z0-9]+/g, '-')
  _str = capitalize(camelize(_str))
  // 不允许数字开头，补充 _
  if (/^\d/.test(_str)) {
    _str = '_' + _str
  }
  return _str
}

export function normalizePagePath(pagePath: string, platform: UniApp.PLATFORM) {
  const absolutePagePath = path.resolve(process.env.UNI_INPUT_DIR, pagePath)
  const isX = process.env.UNI_APP_X === 'true'
  let extensions = isX ? X_PAGE_EXTNAME : PAGE_EXTNAME
  if (platform === 'app') {
    extensions = isX ? X_PAGE_EXTNAME_APP : PAGE_EXTNAME_APP
  }
  for (let i = 0; i < extensions.length; i++) {
    const extname = extensions[i]
    if (fs.existsSync(absolutePagePath + extname)) {
      return pagePath + extname
    }
  }
  console.error(`${pagePath} not found`)
}

export function removeExt(str: string) {
  return str.split('?')[0].replace(/\.\w+$/g, '')
}

const NODE_MODULES_REGEX = /(\.\.\/)?node_modules/g

export function normalizeNodeModules(str: string) {
  str = normalizePath(str).replace(NODE_MODULES_REGEX, 'node-modules')
  // HBuilderX 内置模块路径转换
  str = str.replace(
    /.*\/plugins\/uniapp-cli-vite\/node[-_]modules/,
    'node-modules'
  )
  if (!isInHBuilderX()) {
    // 内部测试
    if (str.includes('uni-app-next/packages/')) {
      str = str.replace(
        /.*\/uni-app-next\/packages\//,
        'node-modules/@dcloudio/'
      )
    }
  }

  if (process.env.UNI_PLATFORM === 'mp-alipay') {
    str = str.replace('node-modules/@', 'node-modules/npm-scope-')
  }
  return str
}

export function normalizeMiniProgramFilename(
  filename: string,
  inputDir?: string
) {
  if (!inputDir || !path.isAbsolute(filename)) {
    return normalizeNodeModules(filename)
  }
  return normalizeNodeModules(path.relative(inputDir, filename))
}

export function normalizeParsePlugins(
  importer: string,
  babelParserPlugins?: ParserPlugin[]
) {
  const isTS = EXTNAME_TS_RE.test(importer.split('?')[0])
  const plugins: ParserPlugin[] = []
  if (isTS) {
    plugins.push('jsx')
  }
  if (babelParserPlugins) plugins.push(...babelParserPlugins)
  if (isTS) plugins.push('typescript', 'decorators-legacy')
  return plugins
}

export function pathToGlob(
  pathString: string,
  glob: string,
  options: { windows?: boolean; escape?: boolean } = {}
): string {
  const isWindows =
    'windows' in options ? options.windows : /^win/.test(process.platform)
  const useEscape = options.escape
  const str = isWindows ? pathString.replace(/\\/g, '/') : pathString
  let safeStr = str.replace(
    /[\\*?[\]{}()!]/g,
    isWindows || !useEscape ? '[$&]' : '\\$&'
  )
  return path.posix.join(safeStr, glob)
}

export function resolveSourceMapPath(
  outputDir?: string,
  platform?: UniApp.PLATFORM
) {
  return path.resolve(
    outputDir || process.env.UNI_OUTPUT_DIR,
    '../.sourcemap/' + (platform || getPlatformDir())
  )
}

function hasProjectYarn(cwd: string) {
  return fs.existsSync(path.join(cwd, 'yarn.lock'))
}

function hasProjectPnpm(cwd: string) {
  return fs.existsSync(path.join(cwd, 'pnpm-lock.yaml'))
}

function getInstallCommand(cwd: string) {
  return hasProjectYarn(cwd)
    ? 'yarn add'
    : hasProjectPnpm(cwd)
    ? 'pnpm i'
    : 'npm i'
}

export function installDepTips(
  type: 'dependencies' | 'devDependencies',
  module: string,
  version?: string
) {
  return `Cannot find module: ${module}
Please run \`${colors.cyan(
    `${getInstallCommand(process.cwd())} ${
      module + (version ? '@' + version : '')
    }${type === 'devDependencies' ? ' -D' : ''}`
  )}\` and try again.`
}

/**
 * 根据路径判断是否为 App.(u?)vue
 * @param {string} filename 相对、绝对路径
 * @returns
 */
export function isAppVue(filename: string) {
  const _filePath = normalizePath(filename)
  return /(\/|\\)app\.(u?)vue$/.test(_filePath.toLowerCase())
}

export function resolveAppVue(inputDir: string) {
  if (process.env.UNI_APP_X === 'true') {
    const appUVue = path.resolve(inputDir, 'App.uvue')
    if (fs.existsSync(appUVue)) {
      return normalizePath(appUVue)
    }
  }
  return normalizePath(path.resolve(inputDir, 'App.vue'))
}

export function parseImporter(importer: string) {
  importer = importer.split('?')[0]
  if (path.isAbsolute(importer)) {
    return normalizePath(path.relative(process.env.UNI_INPUT_DIR, importer))
  }
  return importer
}

export function createResolveErrorMsg(source: string, importer: string) {
  return `Could not resolve "${source}" from "${parseImporter(importer)}"`
}

export function enableSourceMap() {
  if (process.env.UNI_APP_SOURCEMAP === 'true') {
    return true
  }
  if (process.env.UNI_APP_SOURCEMAP === 'false') {
    return false
  }
  return process.env.NODE_ENV === 'development' && isNormalCompileTarget()
}

export function requireUniHelpers() {
  require(path.resolve(
    process.env.UNI_HBUILDERX_PLUGINS,
    'uni_helpers/lib/bytenode'
  ))
  return require(path.join(process.env.UNI_HBUILDERX_PLUGINS, 'uni_helpers'))
}

export function normalizeEmitAssetFileName(fileName: string) {
  const extname = path.extname(fileName)

  if (process.env.UNI_APP_X_TSC === 'true') {
    if (extname !== '.ts') {
      return fileName + '.ts'
    }
  } else {
    // logo.png、pages.json 等
    if (!['.ts', '.uts', '.uvue', '.vue'].includes(extname)) {
      fileName = fileName + '.uts'
    }
  }
  return fileName
}

function createIdent(platform: UniApp.PLATFORM) {
  if (process.env.UNI_INPUT_DIR) {
    const manifestJson = parseManifestJsonOnce(process.env.UNI_INPUT_DIR)
    let id = (manifestJson.appid || '').replace('__UNI__', '')
    const platformAppId = manifestJson[platform]?.appid
    if (platformAppId) {
      id += '%%' + platformAppId
    }
    if (id) {
      return Buffer.from(Buffer.from(id).toString('base64')).toString('hex')
    }
  }
  return ''
}

export function createShadowImageUrl(cdn: number, type: string = 'grey') {
  let identStr = ''
  if (process.env.UNI_PLATFORM !== 'h5' && process.env.UNI_PLATFORM !== 'web') {
    const ident = createIdent(process.env.UNI_PLATFORM)
    identStr = ident ? `${ident}/` : ''
  }
  return `https://cdn${
    (cdn || 0) + (process.env.UNI_APP_X === 'true' ? 1000 : 0) || ''
  }.dcloud.net.cn/${identStr}img/shadow-${type}.png`
}

export function isNormalCompileTarget() {
  // 目前有特殊编译目标 uni_modules 和 ext-api
  return !process.env.UNI_COMPILE_TARGET
}

/**
 * copy from entities 用于转义实体字符
 * @param regex
 * @param map
 * @returns
 */
export function getEscaper(
  regex: RegExp,
  map: Map<number, string>
): (data: string) => string {
  return function escape(data: string): string {
    let match
    let lastIndex = 0
    let result = ''

    while ((match = regex.exec(data))) {
      if (lastIndex !== match.index) {
        result += data.substring(lastIndex, match.index)
      }

      // We know that this character will be in the map.
      result += map.get(match[0].charCodeAt(0))!

      // Every match will be of length 1
      lastIndex = match.index + 1
    }

    return result + data.substring(lastIndex)
  }
}
