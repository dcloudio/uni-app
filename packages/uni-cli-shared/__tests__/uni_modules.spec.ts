import path from 'path'
import {
  findEncryptUniModules,
  findUploadEncryptUniModulesFiles,
  parseUniModulesWithComponents,
} from '../src/uni_modules.cloud'
import { normalizePath } from '../src/utils'

const platforms = ['app-android', 'app-ios', 'web'] as const
describe('uni_modules:uni-ext-api', () => {
  const inputDir = path.resolve(__dirname, '../../playground/uni_modules/src')

  platforms.forEach((platform) => {
    test(`findEncryptUniModules(${platform})`, () => {
      expect(findEncryptUniModules(platform, inputDir)).toMatchSnapshot()
    })
    test(`findUploadEncryptUniModulesFiles(${platform})`, () => {
      const modules = findUploadEncryptUniModulesFiles(
        findEncryptUniModules(platform, inputDir),
        platform,
        inputDir
      )
      expect(
        Object.keys(modules).reduce((res: string[], id: string) => {
          res.push(
            ...modules[id].map((item) => normalizePath(item).split('/src/')[1])
          )
          return res
        }, [])
      ).toMatchSnapshot()
    })
  })
})
describe('uni_modules:cloud', () => {
  const inputDir = path.resolve(__dirname, 'examples/cloud')
  test('parseUniModulesWithComponents', () => {
    expect(
      parseUniModulesWithComponents(inputDir, 'app-android')
    ).toMatchSnapshot()
    expect(parseUniModulesWithComponents(inputDir, 'app-ios')).toMatchSnapshot()
    expect(parseUniModulesWithComponents(inputDir, 'web')).toMatchSnapshot()
    expect(
      parseUniModulesWithComponents(inputDir, 'mp-weixin')
    ).toMatchSnapshot()
  })
})
