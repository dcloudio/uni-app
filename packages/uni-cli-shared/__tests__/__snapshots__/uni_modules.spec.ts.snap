// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uni_modules:cloud parseUniModulesWithComponents 1`] = `
{
  "test-easycom": "
import TestA, { TestAComponentPublicInstance } from './components/test-a/test-a.vue'
import TestB, { TestBComponentPublicInstance } from './components/test-b/test-b.vue'
export { 
  TestA,
  TestAComponentPublicInstance,
  TestB,
  TestBComponentPublicInstance 
}
  ",
  "test-uts1": "export * from './utssdk/index.uts'",
  "test-uts2": "export * from './utssdk/app-android/index.uts'",
}
`;

exports[`uni_modules:cloud parseUniModulesWithComponents 2`] = `
{
  "test-easycom": "
import TestA from './components/test-a/test-a.vue'
import TestB from './components/test-b/test-b.vue'
export { 
  TestA,
  TestB 
}
  ",
  "test-uts1": "export * from './utssdk/index.uts'",
  "test-uts2": "export * from './utssdk/app-ios/index.uts'",
}
`;

exports[`uni_modules:cloud parseUniModulesWithComponents 3`] = `
{
  "test-easycom": "
import TestA from './components/test-a/test-a.vue'
import TestB from './components/test-b/test-b.vue'
export { 
  TestA,
  TestB 
}
  ",
  "test-uts1": "export * from './utssdk/index.uts'",
  "test-uts2": "export * from './utssdk/web/index.uts'",
}
`;

exports[`uni_modules:cloud parseUniModulesWithComponents 4`] = `
{
  "test-easycom": "
import TestA from './components/test-a/test-a.vue'
import TestB from './components/test-b/test-b.vue'
export { 
  TestA,
  TestB 
}
  ",
  "test-uts1": "export * from './utssdk/index.uts'",
  "test-uts2": "export * from './utssdk/mp-weixin/index.uts'",
}
`;

exports[`uni_modules:uni-ext-api findEncryptUniModules(app-android) 1`] = `
{
  "test-com1": undefined,
  "test-com2": undefined,
}
`;

exports[`uni_modules:uni-ext-api findEncryptUniModules(app-ios) 1`] = `
{
  "test-com1": undefined,
  "test-com2": undefined,
}
`;

exports[`uni_modules:uni-ext-api findEncryptUniModules(web) 1`] = `
{
  "test-com1": undefined,
  "test-com2": undefined,
  "test-uts1": undefined,
}
`;

exports[`uni_modules:uni-ext-api findUploadEncryptUniModulesFiles(app-android) 1`] = `
[
  "uni_modules/test-com1/encrypt",
  "uni_modules/test-com1/index.module.uts",
  "uni_modules/test-com1/package.json",
  "uni_modules/test-com1/components/test-com1-1/logo.png",
  "uni_modules/test-com1/components/test-com1-1/test-com1-1.vue",
  "uni_modules/test-com1/components/test-com1-2/test-com1-2.vue",
  "uni_modules/test-com2/encrypt",
  "uni_modules/test-com2/index.module.uts",
  "uni_modules/test-com2/index.uts",
  "uni_modules/test-com2/package.json",
  "uni_modules/test-com2/utils/index.uts",
  "uni_modules/test-com2/components/test-com2-1/api.png",
  "uni_modules/test-com2/components/test-com2-1/test-com2-1.vue",
  "uni_modules/test-com2/components/test-com2-2/test-com2-2.vue",
]
`;

exports[`uni_modules:uni-ext-api findUploadEncryptUniModulesFiles(app-ios) 1`] = `
[
  "uni_modules/test-com1/encrypt",
  "uni_modules/test-com1/index.module.uts",
  "uni_modules/test-com1/package.json",
  "uni_modules/test-com1/components/test-com1-1/test-com1-1.vue",
  "uni_modules/test-com1/components/test-com1-2/test-com1-2.vue",
  "uni_modules/test-com2/encrypt",
  "uni_modules/test-com2/index.module.uts",
  "uni_modules/test-com2/index.uts",
  "uni_modules/test-com2/package.json",
  "uni_modules/test-com2/utils/index.uts",
  "uni_modules/test-com2/components/test-com2-1/test-com2-1.vue",
  "uni_modules/test-com2/components/test-com2-2/test-com2-2.vue",
]
`;

exports[`uni_modules:uni-ext-api findUploadEncryptUniModulesFiles(web) 1`] = `
[
  "uni_modules/test-com1/encrypt",
  "uni_modules/test-com1/index.module.uts",
  "uni_modules/test-com1/package.json",
  "uni_modules/test-com1/components/test-com1-1/test-com1-1.vue",
  "uni_modules/test-com1/components/test-com1-2/test-com1-2.vue",
  "uni_modules/test-com2/encrypt",
  "uni_modules/test-com2/index.module.uts",
  "uni_modules/test-com2/index.uts",
  "uni_modules/test-com2/package.json",
  "uni_modules/test-com2/utils/index.uts",
  "uni_modules/test-com2/components/test-com2-1/test-com2-1.vue",
  "uni_modules/test-com2/components/test-com2-2/test-com2-2.vue",
  "uni_modules/test-uts1/encrypt",
  "uni_modules/test-uts1/package.json",
  "uni_modules/test-uts1/utssdk/index.uts",
]
`;
